tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,239 @@
+    POSIX_ENSURE_REF(msg->msg_iov);
+
+    struct s2n_test_ktls_io_stuffer *io_ctx = (struct s2n_test_ktls_io_stuffer *) conn->send_io_context;
+    POSIX_ENSURE_REF(io_ctx);
+    io_ctx->invoked_count++;
+    /* Assert ancillary_buffer is growable, which simplifies IO logic. Blocking IO can
+     * be mocked by restricting the data_buffer. */
+    POSIX_ENSURE(io_ctx->ancillary_buffer.growable, S2N_ERR_IO);
+
+    size_t total_len = 0;
+    for (size_t count = 0; count < msg->msg_iovlen; count++) {
+        uint8_t *buf = msg->msg_iov[count].iov_base;
+        MOCK_IO_ENSURE_REF(buf);
+        size_t len = msg->msg_iov[count].iov_len;
+
+        /* If we fail to write to stuffer then return blocked */
+        if (s2n_stuffer_write_bytes(&io_ctx->data_buffer, buf, len) < 0) {
+            errno = EAGAIN;
+            return -1;
+        }
+
+        total_len += len;
+    }
+    if (total_len) {
+        /* write record_type and len after data was written successfully. ancillary_buffer is
+         * growable so this operation should always succeed. */
+        MOCK_IO_GUARD(s2n_stuffer_write_uint8(&io_ctx->ancillary_buffer, record_type));
+        MOCK_IO_GUARD(s2n_stuffer_write_uint16(&io_ctx->ancillary_buffer, total_len));
+    }
+
+    return total_len;
+}
+
+/* In userspace TLS, s2n first reads the header to determine the length of next record
+ * and then reads the entire record into conn->in. In kTLS its not possible to know
+ * the length of the next record. Instead the socket returns the minimum of
+ * bytes-requested and data-available; reading multiple consecutive records if they
+ * are of the same type. */
+ssize_t s2n_test_ktls_recvmsg_stuffer_io(struct s2n_connection *conn, struct msghdr *msg, uint8_t *record_type)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->recv_io_context);
+    POSIX_ENSURE_REF(msg);
+    POSIX_ENSURE_REF(msg->msg_iov);
+
+    struct s2n_test_ktls_io_stuffer *io_ctx = (struct s2n_test_ktls_io_stuffer *) conn->recv_io_context;
+    POSIX_ENSURE_REF(io_ctx);
+    /* Assert ancillary_buffer is growable, which simplifies IO logic. Blocking IO can
+     * be mocked by restricting the data_buffer. */
+    POSIX_ENSURE(io_ctx->ancillary_buffer.growable, S2N_ERR_IO);
+    io_ctx->invoked_count++;
+
+    /* s2n only receives using msg_iovlen of 1 */
+    MOCK_IO_ENSURE(msg->msg_iovlen == 1);
+
+    uint8_t *buf = msg->msg_iov->iov_base;
+    MOCK_IO_ENSURE_REF(buf);
+
+    /* There is no data available so return blocked */
+    if (!s2n_stuffer_data_available(&io_ctx->ancillary_buffer)) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    ssize_t total_read = 0;
+    /* updated as partial or multiple records are read */
+    size_t updated_requested_len = msg->msg_iov->iov_len;
+    /* track two record_types since its possible to read multiple records of the same type */
+    *record_type = 0;
+    uint8_t next_record_type = 0;
+    while (*record_type == next_record_type) {
```

Comment:
This is a really strange loop condition, I would assume we should do this read until we run out of room in msg, can you tell me why this is not the case?

Reply:
lets assume we do a read(5) and we have the following two cases below. One will read 5 bytes, while the other will read only 3.
 
`[app|3][app|2]` - reads 5 bytes, both app data records 
`[app|3][alert|2]` - reads 3 bytes, since seconds record is of a different type

Reply:
In the fist iteration both values are 0. In the following iterations we 1) update record_type and 2) attempt to peek for the next record_type. Only if the values are the same then we attempt more reads